migrations/utils.py

```diff
@@ -0,0 +1,167 @@
+    def previous(self, current_node_data) -> object:
+        """Retrieve data before the specified one."""
+        node = self.previous_node(current_node_data)
+        if node:
+            return node.data
+        else:
+            return None
+    
+    def get_sublist(self, first_node_data: str, last_node_data: str = None) -> list:
+        """Return a list consistings of nodes between the specified boundaries.
+        Inclusive of last endpoint, not of first."""
+        unapplied_migrations = []
+        if last_node_data is None:
+            last_node = self._head
+        else:
+            last_node = self.find(last_node_data)
+
+        # Iterate over migrations starting from top
+        while last_node and last_node.data != first_node_data:
+            unapplied_migrations.append(last_node.data)
+            last_node = last_node.prev_node
+
+        # Reverse to account for the order
+        unapplied_migrations.reverse()
+
+        return unapplied_migrations
+
+    def add(self, data):
+        """Add new node after the head. Raises an error if node with such migration already exists."""
+        new_node = self.find(data)
+        if new_node:
+            raise ValueError("adding a duplicate item")
+        else:
+            new_node = Node(data)
+
+            if self._head is None:
+                self._head = new_node
+            else:
+                current_node = self._head
+                while current_node.next_node:
+                    current_node = current_node.next_node
+                current_node.next_node = new_node
+                new_node.prev_node = current_node
+                self._head = new_node
+
+    def build_list_from_dictionary(self, previous_nodes: dict):
+        '''Creates a sorted LinkedList where head is the latest created migration in the directory.
+        Tail is the migration pointing to "None," all new migrations are added after the head'''
+        nodes_references: dict = {}
+        for key in previous_nodes.keys():
+            node = Node(key)
+            nodes_references[key] = node
+
+        # First, create all migration nodes without linking them
+        for migration, node in nodes_references.items():
+            prev_node_id = previous_nodes[migration]
+            if prev_node_id != 'None':
+                prev_node = nodes_references[prev_node_id]
+                if prev_node:
+                    node.prev_node = prev_node
+                    prev_node.next_node = node
+
+        # Find the migration node that has no 'next_node' (i.e., the tail node)
+        for node in nodes_references.values():
+            if node.next_node is None:
+                self._head = node
+                break
+
+        # If no tail node exists and length is not zero, means there is a circual dependency, no outgoing edges
+        if self._head == None:
+            error_message = "Cycle detected in the list"
+            raise ValueError(error_message)
+
+    @property
```

Comment:
@properties are great, but i don't see the value of this one.  aren't you just wrapping what could be done directly?  unless there are additional details handled in the @property or @setter, it's un-pythonic to wrap just for the sake of wrapping.